mountpoint-s3/src/prefetch/part_stream.rs

```diff
@@ -188,18 +191,21 @@ where
 
         let span = debug_span!("prefetch", range=?request_range);
         let client = client.clone();
-        let bucket = bucket.to_owned();
-        let key = key.to_owned();
+        let bucket = config.bucket.clone();
         let task_handle = self
             .runtime
             .spawn_with_handle(
                 async move {
-                    let object_id = ObjectId::new(key, if_match);
                     let (body_sender, body_receiver) = unbounded();
+                    let part_composer = ClientPartComposer {
+                        part_queue_producer,
+                        object_id: config.object_id.clone(),
+                        preferred_part_size: config.preferred_part_size,
+                    };
+
                     let request_reader_future =
-                        read_from_request(body_sender, client, bucket, object_id.clone(), request_range.into());
-                    let part_composer_future =
-                        compose_parts(body_receiver, part_queue_producer, object_id, preferred_part_size);
+                        try_read_from_request(body_sender, client, bucket, config.object_id, request_range.into());
+                    let part_composer_future = part_composer.try_compose_parts(body_receiver);
                     join!(request_reader_future, part_composer_future);
```

Comment:
This `join` over two futures communicating over a channel seems quite convoluted. Have we considered modelling the two components as stream transformer and consumer instead?

Reply:
This is not specific to your changes in this PR, though, so we can also discuss it separately.

Reply:
Here is a throw-away example - just for `read_from_request` - which uses `async_stream`: https://github.com/awslabs/mountpoint-s3/compare/main...passaro:mountpoint-s3:async-stream

Reply:
We haven't considered this option yet. I will have a look at your example, but probably make a separate PR for it if we want to change.

Reply:
Did try `async_stream` and it was really neat! I would make it a followup PR after this one.